### PR TITLE
fix(skate-web/channels): ensure token is authenticated when joining channels

### PR DESCRIPTION
**Pull request by firestack**
# Summary
When initiating a connection to the server the token is validated, but when joining a channel the socket is not checked to be authenticated.

[Phoenix.Channel Documentation on Authorization](https://hexdocs.pm/phoenix/Phoenix.Channel.html#module-authorization) says that 
> Your channels must implement a join/3 callback that authorizes the socket for the given topic.

So this PR modifies all channel `join/3` callbacks to instead be `join_authenticated/3` which is wrapped by `SkateWeb.AuthenticatedChannel` (name and file location not final) so that authentication code does not need to be manually written and managed for new topics, and should make implementing new channels safer.

## Implementation Details
Use of `SkateWeb.AuthenticatedChannel` shims a `Phoenix.Channel` `@callback join/3` into the module, which firsts checks if the socket is authenticated, which delegates to `join_authenticated/3` if the socket token is valid; and returns `{:error, :not_authenticated}` if the token is invalid.
`join_authenticated/3` is defined as a callback matching the `Phoenix.Channel.join/3` spec so that calls made to `join/3` can be transparently passed to `join_authenticated/3`.

## Notes
I've decided to add a macro which inserts a `join/3` callback for `Phoenix.Channel` which wraps calls to `join_authenticated`, the idea being that if you're getting a callback at `join_authenticated` then you should be able to expect that the request has been authenticated already for you.

I'm unsure if this is the way we should do it or if there's a better way I could've approached this, so I'm really interested in feedback.

Adding `SkateWeb.AuthenticatedChannel` to other channels in SkateWeb showed that some tests are also returning false from `valid_token?/1` too soon, but this may need further review.

In addition, I also moved `:valid_token_fn` and `:valid_token?` into the `ChannelAuth` module and consolidated them to a single atom, which can be updated if we prefer a certain atom name.

I've also realized that I could possibly get rid of `AuthenticatedChannel` and instead do the same work within `lib/skate_web.ex`'s `__using__` macro.

# Todo
- [x] Every new thing needs docs
- [ ] Add frontend tests and "response"(refresh)
- [ ] add `AuthenticatedChannel` specific tests
- [ ] Find edge cases to add more elixir tests for
- [ ] (hopefully) find ways to make tests for `join_authenticated` automatically find all the topics that need to be tested.
- [x] Change elixir test names to be more descriptive 

---

Asana Ticket: https://app.asana.com/0/0/1203412139193187

### CODE REVIEW

File: lib/skate_web/channels/alerts_channel.ex [hunk 1]
```diff
@@ -4,18 +4,16 @@ defmodule SkateWeb.AlertsChannel do
...
-  def join("alerts:route:" <> route_id, _message, socket) do
+  use SkateWeb.AuthenticatedChannel
+  @impl SkateWeb.AuthenticatedChannel
+  def join_authenticated("alerts:route:" <> route_id, _message, socket) do
     alerts = Duration.log_duration(Server, :subscribe_to_alerts, [route_id])
     {:ok, %{data: alerts}, socket}
   end
 
   @impl Phoenix.Channel
   def handle_info({:new_realtime_data, lookup_args}, socket) do
```
Comment (lemald) [inline lib/skate_web/channels/alerts_channel.ex:15]: **thought:** I wonder if we could similarly wrap `handle_info` and also get rid of some code duplication there...

File: lib/skate_web/channels/channel_auth.ex [hunk 2]
```diff
@@ -6,6 +6,12 @@ defmodule SkateWeb.ChannelAuth do
   """
   @spec valid_token?(Phoenix.Socket.t()) :: boolean()
   def valid_token?(socket) do
+    token_fn = Application.get_env(:skate, :valid_token_fn, &_valid_token?/1)
+    token_fn.(socket)
+  end
+
+  @spec _valid_token?(Phoenix.Socket.t()) :: boolean()
+  defp _valid_token?(socket) do
```
Comment (lemald) [inline lib/skate_web/channels/channel_auth.ex:14]: **nitpick:** I know in Elixir, prefixing a variable with `_` causes the compiler to not throw a warning if it isn't used. I'm trying to make up my mind as to whether using it in this context is confusing or not, but I also don't have a better idea. Maybe `default_valid_token?` or something?

File: lib/skate_web/channels/notifications_channel.ex [hunk 3]
```diff
@@ -15,8 +12,10 @@ defmodule SkateWeb.NotificationsChannel do
     end
   end
 
-  @impl true
-  def join("notifications", _message, socket) do
+  use SkateWeb.AuthenticatedChannel
```
Comment (lemald) [inline lib/skate_web/channels/notifications_channel.ex:15]: **nitpick:** I guess this isn't a strict requirement since it passes CI, but I think it would be easier to read this file if all of the `use` / `import` / `alias` statements are grouped together up top.

File: lib/skate_web/channels/channel_auth.ex (hunk 2 above)
Comment (firestack) [reply lib/skate_web/channels/channel_auth.ex:14]: Oh right, I don't wanna keep the leading underscore, I'll think on it!

File: test/skate_web/channels/alerts_channel_test.exs [hunk 4]
```diff
@@ -24,6 +24,18 @@ defmodule SkateWeb.AlertsChannelTest do
...
+      reassign_env(:skate, :valid_token_fn, fn _socket -> false end)
+
+      for %{result: result, route: route} <- [
+            %{result: {:error, :not_authenticated}, route: "alerts:route:"},
+            %{result: {:error, :not_authenticated}, route: "vehicles:route:"},
+            %{result: {:error, :not_authenticated}, route: "random:topic:1"},
+            %{result: {:error, :not_authenticated}, route: "random:topic:2"}
+          ],
+          do: assert(^result = subscribe_and_join(socket, AlertsChannel, route))
+    end
```
Comment (lemald) [inline test/skate_web/channels/alerts_channel_test.exs:38]: **suggestion:** it's a little confusing to have this specifically in the `AlertsChannel` test since it tests several different channels. One option is to simply make this a bunch of different tests in different channel modules, or maybe there's some good unified place to put this one test that makes sense.

File: lib/skate_web/channels/channel_auth.ex (hunk 2 above)
Comment (firestack) [reply lib/skate_web/channels/channel_auth.ex:14]: @lemald what do you think of `valid_socket_token?`?

File: test/skate_web/channels/alerts_channel_test.exs (hunk 4 above)
Comment (firestack) [reply test/skate_web/channels/alerts_channel_test.exs:38]: Oh that reminds me I need to go over and change all the test names, since I just copy pasted.

File: lib/skate_web/channels/channel_auth.ex (hunk 2 above)
Comment (lemald) [reply lib/skate_web/channels/channel_auth.ex:14]: Sure, that works.

File: test/skate_web/channels/alerts_channel_test.exs (hunk 4 above)
Comment (firestack) [reply test/skate_web/channels/alerts_channel_test.exs:38]: also this is only testing `AlertsChannel`, I was leaving random topics in there when testing to ensure I was hitting strange edge cases should they appear.

**Comment by lemald** [comment]
@firestack looks like there are a few things with the tests that you're still working on, but overall I like the direction this is headed!

File: lib/skate_web/channels/channel_auth.ex (hunk 2 above)
Comment (firestack) [reply lib/skate_web/channels/channel_auth.ex]: interesting note
https://hexdocs.pm/elixir/writing-documentation.html#hiding-internal-modules-and-functions
> Start the function name with one or two underscores, for example, `__add__/2`. Functions starting with underscore are automatically treated as hidden, although you can also be explicit and add `@doc false`. The compiler does not import functions with leading underscores and they hint to anyone reading the code of their intended private usage.

Comment (lemald) [reply lib/skate_web/channels/channel_auth.ex]: Interesting - usually we just use `defp` for private functions and leave it at that, but it's good to know that at least some parts of the Elixir documentation recommend the underscore approach.

**Review by firestack — COMMENTED** [review]
I added some ideas I had that felt like bikeshedding, I left them as review comments, let me know if you like them and I'll integrate them.

I still haven't figured out how to let the frontend know that the channel validity has expired, if @mbta/skate-developers have any ideas on how to implement/handle that let me know, I'm not even sure how to test this all the way through the frontend to the backend.

I think the tests I added are a bit to repetitiive, and they might benifit from being consolodated into a `AuthenticatedChannel.test.exs` instead of `describe("join/3")` since we're just enumerating topics. But I wanted the tests to stay local so they're easy to find. Any feedback welcome.

I had an idea that maybe we could scrape the `@spec`s and generate tests from the topics but that's probably it's own ticket.

File: lib/skate_web/channels/channel_auth.ex [hunk 5]
```diff
@@ -6,14 +6,14 @@ defmodule SkateWeb.ChannelAuth do
...
-      {:ok, _claims} ->
-        true
+  @spec valid_socket_token?(Phoenix.Socket.t()) :: boolean()
+  defp valid_socket_token?(socket) do
+    token = Guardian.Phoenix.Socket.current_token(socket)
 
-      _ ->
-        false
-    end
+    match?({:ok, _claims}, AuthManager.decode_and_verify(token))
```
[the remark this answers is not in the document]
Comment (lemald) [reply lib/skate_web/channels/channel_auth.ex:17]: I always like rewriting things as pipelines so I'd say to go ahead and integrate this change. 👍

**Comment by firestack** [comment]
I'm gonna merge this and see if we get any logs on dev.
